lib/clib/Corres_UL_C.thy

```diff
@@ -288,6 +288,43 @@ lemma ccorres_from_vcg_nofail:
   apply (rule hoare_complete, simp add: HoarePartialDef.valid_def)
   done
 
+lemma ccorres_to_vcg_with_prop:
+  "\<lbrakk>ccorres_underlying srel \<Gamma> rrel xf arrel axf P P' [] a c; no_fail Q a; \<lbrace>P\<rbrace> a \<lbrace>R\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<Gamma> \<turnstile> {s'. P s \<and> Q s \<and> s' \<in> P' \<and> (s, s') \<in> srel}
+           c {t'. \<exists>(rv, t) \<in> fst (a s). (t, t') \<in> srel \<and> rrel rv (xf t') \<and> R rv t}"
+  apply (clarsimp simp: no_fail_def)
+  apply (drule_tac x=s in spec)
+  apply (rule hoare_complete)
+  apply (simp add: HoarePartialDef.valid_def cvalid_def)
+  apply (intro impI allI)
+  apply clarsimp
+  apply (frule (5) ccorres_empty_handler_abrupt)
+  apply (erule (4) ccorresE)
+   apply (erule (1) EHOther)
+  apply (rule image_eqI)
+   apply simp
+  apply (fastforce simp: unif_rrel_simps valid_def)
+  done
+
+lemma ccorres_to_vcg_with_prop':
+  "\<lbrakk>ccorres_underlying srel \<Gamma> rrel xf arrel axf P P' [] a c; no_fail Q a; \<lbrace>P\<rbrace> a \<lbrace>R\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<Gamma> \<turnstile> {s'. P s \<and> Q s \<and> s' \<in> P' \<and> (s, s') \<in> srel}
+           c {t'. \<exists>rv t. (t, t') \<in> srel \<and> rrel rv (xf t') \<and> R rv t}"
+  apply (frule (2) ccorres_to_vcg_with_prop[where s=s])
+  apply (fastforce elim: conseq_under_new_pre)
+  done
```

Comment:
I'd be left scratching my head as to when to use this rule vs any of the other ones.
Suggestion: since it's a two-line proof without interesting bits, use a `by`

Reply:
I might be able to get rid of this one, actually. I can use the stronger form just above, I'm pretty sure.

The `with_prop` versions are probably useful only in some sort of a `ccorres` rule for some sort of loop, where we'd like the `\exists rv t` bit and then want to carry a loop invariant through, too. But you get the property almost for free, so I thought I could prove this and then get `ccorres_to_vcg` from having the trivial postcondition

Reply:
I was able to get rid of this rule. So now there's just a question of whether to clean up the other variants of this rule, `ccorres_to_vcg'` and `ccorres_to_vcg_nf`

Reply:
Cool. I'd say yes, if any rules become easily provable due to the new one you've added, then cleaning up their proofs is preferable.